tutorial/markdown/python/python-langchain-pdf-chat/python-langchain-pdf-chat.md

```diff
@@ -0,0 +1,584 @@
+
+```python
+retriever = vector_store.as_retriever()
+```
+
+### Prompt Chain
+
+A prompt for a language model is a set of instructions or input provided by a user to guide the model's response, helping it understand the context and generate relevant and coherent language-based output, such as answering questions, completing sentences, or engaging in a conversation. We will use a template and create a [prompt chain](https://python.langchain.com/docs/modules/model_io/prompts/quick_start/) using [_ChatPromptTemplate_](https://python.langchain.com/docs/modules/model_io/prompts/quick_start/#chatprompttemplate) Class of LangChain
+
+```python
+template = """You are a helpful bot. If you cannot answer based on the context provided, respond with a generic answer. Answer the question as truthfully as possible using the context below:
+{context}
+
+Question: {question}"""
+
+prompt = ChatPromptTemplate.from_template(template)
+```
+
+### LLM Chain
+
+Large Language Models (LLMs) are a core component of LangChain. LangChain does not serve its own LLMs, but rather provides a standard interface for interacting with many LLMs. To be specific, this interface is one that takes as input a string and returns a string. We will use [ChatOpenAI](https://python.langchain.com/docs/integrations/chat/openai/) LLM Model. We can also set other parameters like model, API_KEY, temperature to be used for this model.
+
+```python
+# Use OpenAI GPT 4 as the LLM for the RAG
+llm = ChatOpenAI(temperature=0, model="gpt-4-1106-preview", streaming=True)
+```
+
+### Combining to a single chain
+
+We can combine these different modules to a single chain which will run synchronously. The pipe operator (|) is used to connect these components, forming a chain of operations.
+
+The input dictionary with the "context" and "question" keys is passed through each component. The retriever component retrieves relevant information from the vector store, which is then combined with the [question](https://python.langchain.com/docs/expression_language/primitives/passthrough/) to generate a prompt. This prompt is passed to the language model (llm), which generates a response based on both the question and the retrieved context. Finally, the output is parsed into a string by [StrOutputParser()](https://python.langchain.com/docs/modules/model_io/concepts/#stroutputparser).
+
+```python
+# RAG chain
+chain = (
+    {"context": retriever, "question": RunnablePassthrough()}
+    | prompt
+    | llm
+    | StrOutputParser()
+)
+```
+
+### Chain without RAG
+
+We will repeat the same Process as above however this will not be context aware from RAG. Basically we will directly call LLM from the user question. Basically, every step is same just we will not use retriever.
+
+```python
+template_without_rag = """You are a helpful bot. Answer the question as truthfully as possible.
+
```

Comment:
```suggestion
We will repeat the same process as above however this will not have the context from the vector store. Basically we will directly call LLM from the user question. Basically, every step is same just that we will not use retriever.
```